src/ReverseProxy/LoadBalancing/SlowStartDestinationSelector.cs

```diff
@@ -0,0 +1,88 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using Microsoft.Extensions.Logging;
+using Yarp.ReverseProxy.Model;
+using Yarp.ReverseProxy.Utilities;
+
+namespace Yarp.ReverseProxy.LoadBalancing;
+
+internal sealed class SlowStartDestinationSelector
+{
+    private const int SlowStartWindowSeconds = 200; // TODO configurable
+
+    private readonly IRandomFactory _randomFactory;
+    private readonly ILogger _logger;
+
+    public SlowStartDestinationSelector(IRandomFactory randomFactory, ILogger logger)
+    {
+        _randomFactory = randomFactory;
+        _logger = logger;
+    }
+
+    public DestinationState? PickRandomDestination(IReadOnlyList<DestinationState> availableDestinations)
+    {
+        if (availableDestinations.Count == 0)
+        {
+            return null;
+        }
+        var computedWeights = ComputeWeights(availableDestinations);
+
+        var destinationIndex = PickRandomDestination(computedWeights);
+        return availableDestinations[destinationIndex];
+    }
+
+    public int PickRandomDestination((List<(DestinationState destination, int weight)> destinationWeights, int totalWeight) computedWeights)
+    {
+        var destinationWeights = computedWeights.destinationWeights;
+        var totalWeight = computedWeights.totalWeight;
+
+        var random = _randomFactory.CreateRandomInstance();
+        var chosenInstance = random.Next(totalWeight);
+
+        _logger.LogInformation("Picking destination {0} from {1}. TotalWeight={2}", chosenInstance, destinationWeights.Select(x => (x.destination.DestinationId, x.weight)).ToArray(), totalWeight);
+
+        for (var i = 0; i < destinationWeights.Count - 1; i++)
+        {
```

Comment:
Isn't destinationWeights supposed to be sorted in descendinging order of weights?